ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl/Pure.hs

```diff
@@ -0,0 +1,343 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | Pure side of the Mempool implementation.
+--
+-- Operations are performed in a pure style returning data types that model
+-- the control flow through the operation and can then be interpreted to perform
+-- the actual STM/IO operations.
+
+module Ouroboros.Consensus.Mempool.Impl.Pure (
+    -- * Mempool
+    pureGetSnapshotFor
+  , pureRemoveTxs
+  , pureSyncWithLedger
+  , pureTryAddTxs
+  , runRemoveTxs
+  , runSyncWithLedger
+  , runTryAddTxs
+    -- * MempoolSnapshot
+  , implSnapshotFromIS
+  ) where
+
+import           Control.Exception (assert)
+import           Data.Maybe (isJust, isNothing)
+import qualified Data.Set as Set
+import           Data.Word (Word32)
+
+import           Control.Tracer
+import           Ouroboros.Consensus.HeaderValidation
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.Ledger.SupportsMempool
+import           Ouroboros.Consensus.Mempool.API
+import           Ouroboros.Consensus.Mempool.Impl.Types
+import           Ouroboros.Consensus.Mempool.TxSeq (TicketNo, TxTicket (..),
+                     zeroTicketNo)
+import qualified Ouroboros.Consensus.Mempool.TxSeq as TxSeq
+import           Ouroboros.Consensus.Util (whenJust)
+import           Ouroboros.Consensus.Util.IOLike
+
+{-------------------------------------------------------------------------------
+  Mempool Implementation
+-------------------------------------------------------------------------------}
+
+-- | Data constructor containing the control flow for 'tryAddTxs'. Can be
+-- interpreted in an IOLike monad with 'runTryAddTxs'.
+data TryAddTxs blk =
+    -- | No more transactions to add or the mempool reached its full capacity.
+    Done [GenTx blk]
+    -- | There are more transactions to process.
+  | More
+      (Maybe (InternalState blk))
+      -- ^ If the transaction was accepted, the new state that can be written to
+      -- the TVar.
+      (TraceEventMempool blk)
+      -- ^ The event emitted by the operation.
+      (MempoolAddTxResult blk)
+      -- ^ The result of trying to add the transaction to the mempool.
+      (InternalState blk -> TryAddTxs blk)
+      -- ^ The continuation.
+
+-- | Add a list of transactions (oldest to newest) by interpreting a 'TryAddTxs'
+-- from 'pureTryAddTxs'.
+--
+-- This function returns two lists: the transactions that were added or
+-- rejected, and the transactions that could not yet be added, because the
+-- Mempool capacity was reached. See 'addTxs' for a function that blocks in
+-- case the Mempool capacity is reached.
+--
+-- Transactions are added one by one, updating the Mempool each time one was
+-- added successfully.
+--
+-- This function does not sync the Mempool contents with the ledger state in
+-- case the latter changes, it relies on the background thread to do that.
+--
+-- POSTCONDITON:
```

Comment:
Javier noticed that this comment has been out-of-date since my Mempool changes for Alonzo. He's going to simplify this comment, saying something like `See the necessary invariants on the Haddock for 'API.tryAddTxs'.`

Reply:
Added a reference to the invariants in both comments that needed it.